include/etl/iterator.h

```diff
@@ -604,7 +604,191 @@ namespace etl
     return etl::move_iterator<TIterator>(itr);
   }
 
-#endif
+#endif //ETL_USING_CPP11
+
+  //***************************************************************************
+  // back_insert_iterator
+  //***************************************************************************
+
+  /**
+   * @brief  Turns assignment into insertion.
+   *
+   * These are output iterators, constructed from a container-of-T.
+   * Assigning a T to the iterator appends it to the container using push_back.
+   *
+   * @tparam TContainer
+   */
+  template <class TContainer>
+  class back_insert_iterator : public iterator<output_iterator_tag, void, void, void, void>
+  {
+  public:
+    /// A nested typedef for the type of whatever container you used.
+    typedef TContainer container_type;
+
+    /// The only way to create this %iterator is with a container.
+    explicit ETL_CONSTEXPR17 back_insert_iterator(TContainer& c)
+      : container(addressof(c))
+    {
+    }
+
+    /**
+     * This kind of %iterator doesn't really have a @a position in the
+     * container (you can think of the position as being permanently at
+     * the end, if you like).  Assigning a value to the %iterator will
+     * always append the value to the end of the container.
+     *
+     * @param value An instance of whatever type container_type::const_reference is;
+     *              presumably a reference-to-const T for container<T>.
+     * @return This %iterator, for chained operations.
+     */
+    ETL_CONSTEXPR17 back_insert_iterator& operator=(const typename TContainer::value_type& value)
```

Comment:
I have used `ETL_CONSTEXPR17` but I just found there is also `ETL_CONSTEXPR20`, `ETL_CONSTEXPR`, `ETL_IF_CONSTEXPR`.
Are there some guidelines to help select the correct one ?

Reply:
The `ETL_CONSTEXPRxx` macros are defined to `constexpr` or 'nothing' depending on the C++ language level.

For example `ETL_CONSTEXPR14` is only defined as `constexpr` for C++14 and above.

`ETL_IF_CONSTEXPR` is defined as `constexpr` for C++17 and above.
It is used in this context...
`if ETL_IF_CONSTEXPR(etl::is_signed_v<T>)`

Reply:
`ETL_CONSTEXPR` >= C++11
`ETL_CONSTEXPR14` >= C++14
`ETL_CONSTEXPR17` >= C++17
`ETL_CONSTEXPR20` >= C++20
`ETL_IF_CONSTEXPR` >= C++17

Reply:
A big difference between C++11 and C++14 was to do with what was allowed in a `constexpr` function.
I try to remember to declare new additions to the ETL as `constexpr` wherever I can.
If member functions in a class could be a mixture or `ETL_CONSTEXPR` or `ETL_CONSTEXPR14` I tend to stick to `ETL_CONSTEXPR14` for all of them.